ocean_lib/models/v4/test/test_factory_router.py

```diff
@@ -0,0 +1,543 @@
+    """Tests that if it fails to add a new poolTemplate if NOT Router Owner"""
+    factory_router = FactoryRouter(
+        web3,
+        get_factory_router_address(config),
+    )
+    with pytest.raises(exceptions.ContractLogicError) as err:
+        factory_router.add_pool_template(
+            get_another_consumer_wallet().address, get_consumer_wallet()
+        )
+    assert (
+        err.value.args[0]
+        == "execution reverted: VM Exception while processing transaction: revert OceanRouter: NOT OWNER"
+    )
+    assert (
+        factory_router.is_pool_template(get_another_consumer_wallet().address) == False
+    )
+
+
+def test_add_pool_template(web3, config):
+    """Tests that poolTemplate is added if Router owner"""
+    consumer = get_consumer_wallet().address
+    factory_router = FactoryRouter(
+        web3,
+        get_factory_router_address(config),
+    )
+    factory_router.remove_pool_template(
+        consumer, get_factory_deployer_wallet(network=_NETWORK)
+    )
+    assert factory_router.is_pool_template(consumer) == False
+    factory_router.add_pool_template(
+        consumer, get_factory_deployer_wallet(network=_NETWORK)
+    )
+    assert factory_router.is_pool_template(consumer) == True
+
+
+def test_remove_pool_template(web3, config):
+    """Tests that poolTemplate is removed if Router owner"""
+    consumer = get_consumer_wallet().address
+    factory_router = FactoryRouter(
+        web3,
+        get_factory_router_address(config),
+    )
+    assert factory_router.is_pool_template(consumer) == True
+    factory_router.remove_pool_template(
+        consumer, get_factory_deployer_wallet(network=_NETWORK)
+    )
+    assert factory_router.is_pool_template(consumer) == False
+
+
+def test_fail_remove_pool_template(web3, config):
+    """Tests that if it fails to remove a poolTemplate if NOT Router Owner"""
+    consumer = get_consumer_wallet().address
+    factory_router = FactoryRouter(
+        web3,
+        get_factory_router_address(config),
+    )
+
+    factory_router.add_pool_template(
+        consumer, get_factory_deployer_wallet(network=_NETWORK)
+    )
+    assert factory_router.is_pool_template(consumer) == True
+
+    with pytest.raises(exceptions.ContractLogicError) as err:
+        factory_router.remove_pool_template(consumer, get_consumer_wallet())
+
+    assert (
+        err.value.args[0]
+        == "execution reverted: VM Exception while processing transaction: revert OceanRouter: NOT OWNER"
+    )
+    assert factory_router.is_pool_template(consumer) == True
+
+
+def test_buy_dt_batch(web3, config):
+    """Tests that a batch of tokens is successfully bought through the buy_dt_batch function"""
+    factory_router = FactoryRouter(
+        web3,
+        get_factory_router_address(config),
+    )
+
+    nft_factory = ERC721FactoryContract(
+        web3=web3, address=get_nft_factory_address(config)
+    )
+
+    consumer_wallet = get_consumer_wallet()
```

Comment:
this should already be a fixture. If not, please add it to conftest. But check in with @maldo-ctrl as well, since I reviewed the same thing on his PR. :)